File: road/serde/packet_tools.go
package serde

import (
	"github.com/lixianmin/got/iox"
	"io"
)

/********************************************************************
created:    2023-06-05
author:     lixianmin

Copyright (C) - All Rights Reserved
*********************************************************************/

func EncodePacket(writer *iox.OctetsWriter, pack Packet) {
	_ = writer.Write7BitEncodedInt(pack.Kind)

	// pack.Kind是kind还是RouteBase+len(route), 这是在调用EncodePacket之前就准备好的
	if pack.Kind > RouteBase {
		_ = writer.Stream().Write(pack.Route)
	}

	_ = writer.Write7BitEncodedInt(pack.RequestId)
	_ = writer.WriteBytes(pack.Code)
	_ = writer.WriteBytes(pack.Data)
}

func DecodePacket(reader *iox.OctetsReader) ([]Packet, error) {
	var packets []Packet = nil
	var stream = reader.Stream()

	for {
		var lastPosition = stream.Position()

		var kind, err = reader.Read7BitEncodedInt()
		if err == iox.ErrNotEnoughData {
			rewindStream(stream, lastPosition)
			return packets, nil
		}

		var route []byte = nil
		if kind > RouteBase {
			var size = kind - RouteBase
			var data = make([]byte, size)
			var num, err2 = stream.Read(data)
			if err2 == iox.ErrNotEnoughData || num != int(size) {
				rewindStream(stream, lastPosition)
				return packets, nil
			}

			route = data
		}

		requestId, err := reader.Read7BitEncodedInt()
		if err == iox.ErrNotEnoughData {
			rewindStream(stream, lastPosition)
			return packets, nil
		}

		code, err := reader.ReadBytes()
		if err == iox.ErrNotEnoughData {
			rewindStream(stream, lastPosition)
			return packets, nil
		}

		data, err := reader.ReadBytes()
		if err == iox.ErrNotEnoughData {
			rewindStream(stream, lastPosition)
			return packets, nil
		}

		var pack = Packet{Kind: kind, Route: route, RequestId: requestId, Code: code, Data: data}
		packets = append(packets, pack)
	}
}

func rewindStream(stream *iox.OctetsStream, position int) {
	_, _ = stream.Seek(int64(position), io.SeekStart)
}

File: web/src/code/road/packet_tools.js
'use strict'

/********************************************************************
 created:    2023-06-09
 author:     lixianmin

 Copyright (C) - All Rights Reserved
 *********************************************************************/
import {SeekOrigin} from "@src/code/iox/octets_stream";
import {PacketKind} from "@src/code/road/consts";

export function encode(writer, pack) {
    writer.write7BitEncodedInt(pack.kind)

    // pack.kind是kind还是RouteBase+len(route), 这是在调用EncodePacket之前就准备好的
    // route需要是一个[]byte而不能是string, 因为需要在外围计算route的长度, 并赋值到kind, 计算前我们就已经拿到[]byte了
    if (pack.kind > PacketKind.RouteBase) {
        writer.stream.write(pack.route, 0, pack.route.length)
    }

    writer.write7BitEncodedInt(pack.requestId)
    writer.writeBytes(pack.code)
    writer.writeBytes(pack.data)
}

export function decode(reader) {
    const packets = []
    const stream = reader.stream

    while (true) {
        const lastPosition = stream.position
        try {
            const kind = reader.read7BitEncodedInt()

            let route = undefined
            if (kind > PacketKind.RouteBase) {
                const size = kind - PacketKind.RouteBase
                route = new Uint8Array(size)
                stream.read(route, 0, size)
            }

            const requestId = reader.read7BitEncodedInt()
            const code = reader.readBytes()
            const data = reader.readBytes()

            const pack = {kind, route, requestId, code, data}
            packets.push(pack)
        } catch (ex) {
            // console.log(ex, stream)
            stream.seek(lastPosition, SeekOrigin.Begin)
            return packets
        }
    }
}
